packages/xlucene-evaluator/src/parser/context.ts

```diff
@@ -81,10 +89,30 @@ export default function makeContext(args: any) {
     }
 
     function parseFunction(field: string, name: string, params: i.Term[]) {
-        // TODO: get better types name
         const fnType = xluceneFunctions[name];
         if (fnType == null) throw new Error(`Could not find an xlucene function with name "${name}"`);
-        return fnType.create(field, params, { logger, typeConfig });
+        // we are delaying instantiation until after parser since this can be called multiple times
+        return () => fnType.create(field, params, { logger, typeConfig });
+    }
+
+    function makeFlow(field: string, values: any[]) {
+        return values.map((value) => {
+            validateRestrictedVariable(value);
+            const node = { field, type: i.ASTType.Term, value };
+            coerceTermType(node);
+            // this creates an OR statement
+            return {
+                type: i.ASTType.Conjunction,
+                nodes: [node]
+            };
+        });
+    }
+    // cannot allow variables that are objects, errors, maps, sets, buffers
+    function validateRestrictedVariable(val: any) {
+        const type = typeof val;
+        if (!['string', 'number', 'boolean'].includes(type)) {
+            throw new Error(' non-function variables may only be numbers, strings, boolean or an array of these primitives');
```

Comment:
This error message isn't very human friendly. I think something like `Unsupported type of Buffer  received for variable $foo` instead of ` non-function variables may only be numbers, strings, boolean or an array of these primitives`

```suggestion
            throw new Error('Unsupported type of ${getTypeOf(val)} received for variable $${variable}');
```